### PR TITLE
Added missing type annotations for fractions and numbers modules.

### PR DESCRIPTION
**Pull request by erictraut**

### CODE REVIEW

File: stdlib/2and3/fractions.pyi [hunk 1]
```diff
@@ -39,33 +39,33 @@ class Fraction(Rational):
...
     def denominator(self) -> int: ...
-    def __add__(self, other): ...
-    def __radd__(self, other): ...
-    def __sub__(self, other): ...
-    def __rsub__(self, other): ...
-    def __mul__(self, other): ...
-    def __rmul__(self, other): ...
-    def __truediv__(self, other): ...
-    def __rtruediv__(self, other): ...
+    def __add__(self, other: _ComparableNum) -> Fraction: ...
```
Comment (JelleZijlstra) [inline stdlib/2and3/fractions.pyi:42]: This isn't true:
```
In [49]: Fraction(1, 2) + pi
Out[49]: 3.641592653589793
```
And `Fraction(1, 2) + Decimal(5)` throws; it looks like it just supports int, float, Fraction, and complex as the argument type.

Maybe we can do this correctly with overloads, but it's also fine to just leave the return types for most of these as Any.

File: stdlib/2and3/fractions.pyi [hunk 2]
```diff
@@ -39,33 +39,33 @@ class Fraction(Rational):
...
+    def __rpow__(self, other: _ComparableNum) -> Fraction: ...
     def __pos__(self) -> Fraction: ...
     def __neg__(self) -> Fraction: ...
     def __abs__(self) -> Fraction: ...
     def __trunc__(self) -> int: ...
     if sys.version_info >= (3, 0):
         def __floor__(self) -> int: ...
         def __ceil__(self) -> int: ...
-        def __round__(self, ndigits: Optional[Any] = ...): ...
+        def __round__(self, ndigits: Optional[Any] = ...) -> Fraction: ...
```
Comment (JelleZijlstra) [inline stdlib/2and3/fractions.pyi:68]: If ndigits is None (the default) this actually returns an int.

File: stdlib/2and3/fractions.pyi [hunk 3]
```diff
@@ -83,3 +83,4 @@ class Fraction(Rational):
     @property
     def imag(self) -> Fraction: ...
     def conjugate(self) -> Fraction: ...
+
```
Comment (JelleZijlstra) [inline stdlib/2and3/fractions.pyi]: Remove this, it's causing CI failures

File: stdlib/2and3/numbers.pyi [hunk 4]
```diff
@@ -22,42 +24,42 @@ class Complex(Number):
         def __nonzero__(self) -> bool: ...
     @property
     @abstractmethod
-    def real(self): ...
+    def real(self: _T) -> _T: ...
```
Comment (JelleZijlstra) [inline stdlib/2and3/numbers.pyi:27]: That doesn't seem right, the real part of a complex number is not itself a complex number. Documentation suggests this should return a Real.

File: stdlib/2and3/numbers.pyi [hunk 5]
```diff
@@ -22,42 +24,42 @@ class Complex(Number):
...
     @abstractmethod
-    def __pow__(self, exponent): ...
+    def __pow__(self: _T, exponent: Any) -> _T: ...
     @abstractmethod
-    def __rpow__(self, base): ...
-    def __abs__(self): ...
-    def conjugate(self): ...
-    def __eq__(self, other: object) -> bool: ...
+    def __rpow__(self: _T, base: Any) -> _T: ...
+    def __abs__(self: _T) -> _T: ...
```
Comment (JelleZijlstra) [inline stdlib/2and3/numbers.pyi:58]: The absolute value of a complex number isn't a complex, it's a real. For builtins.complex it returns a float.

File: stdlib/2and3/numbers.pyi [hunk 6]
```diff
@@ -70,31 +72,27 @@ class Real(Complex, SupportsFloat):
...
         def __ceil__(self) -> int: ...
         @abstractmethod
-        @overload
-        def __round__(self, ndigits: None = ...): ...
-        @abstractmethod
-        @overload
-        def __round__(self, ndigits: int): ...
-    def __divmod__(self, other): ...
-    def __rdivmod__(self, other): ...
+        def __round__(self: _T, ndigits: Optional[int] = ...) -> _T: ...
```
Comment (JelleZijlstra) [inline stdlib/2and3/numbers.pyi:75]: The docstring says "If ndigits is omitted or None, returns an Integral, otherwise returns a Real".

**Comment by erictraut** [comment]
@JelleZijlstra thanks for reviewing! I think I've addressed all of your feedback. Let me know if there's anything remaining.

File: stdlib/2and3/numbers.pyi [hunk 7]
```diff
@@ -22,42 +24,42 @@ class Complex(Number):
...
     @abstractmethod
-    def real(self): ...
+    def real(self) -> Any: ...
     @property
     @abstractmethod
-    def imag(self): ...
+    def imag(self) -> Any: ...
     @abstractmethod
-    def __add__(self, other): ...
+    def __add__(self: _T, other: Any) -> _T: ...
```
Comment (JelleZijlstra) [inline stdlib/2and3/numbers.pyi:32]: This isn't necessarily true; many concrete number classes return a different type from `__add__`. For example, int + float gives you a float. Unfortunately most of the methods on Number and Real probably have to just return Any.

File: stdlib/2and3/fractions.pyi [hunk 8]
```diff
@@ -39,33 +40,88 @@ class Fraction(Rational):
...
-    def __add__(self, other: _ComparableNum) -> Fraction: ...
-    def __radd__(self, other: _ComparableNum) -> Fraction: ...
-    def __sub__(self, other: _ComparableNum) -> Fraction: ...
-    def __rsub__(self, other: _ComparableNum) -> Fraction: ...
-    def __mul__(self, other: _ComparableNum) -> Fraction: ...
-    def __rmul__(self, other: _ComparableNum) -> Fraction: ...
-    def __truediv__(self, other: _ComparableNum) -> Fraction: ...
-    def __rtruediv__(self, other: _ComparableNum) -> Fraction: ...
+    @overload
+    def __add__(self, other: Union[int, Fraction]) -> Fraction: ...
```
[the remark this answers is not in the document]
Comment (erictraut) [reply stdlib/2and3/fractions.pyi:44]: Good catch. I've added overloads for complex parameters.

File: stdlib/2and3/fractions.pyi [hunk 9]
```diff
@@ -39,33 +40,88 @@ class Fraction(Rational):
...
     def __pos__(self) -> Fraction: ...
     def __neg__(self) -> Fraction: ...
     def __abs__(self) -> Fraction: ...
     def __trunc__(self) -> int: ...
     if sys.version_info >= (3, 0):
         def __floor__(self) -> int: ...
         def __ceil__(self) -> int: ...
-        def __round__(self, ndigits: Optional[Any] = ...) -> Fraction: ...
+        @overload
+        def __round__(self) -> int: ...
```
[the remark this answers is not in the document]
Comment (erictraut) [reply stdlib/2and3/fractions.pyi]: Done

File: stdlib/2and3/numbers.pyi (hunk 7 above)
Comment (erictraut) [reply stdlib/2and3/numbers.pyi]: Yeah, that is unfortunate. I've removed all use of _T in this file and switched everything to Any.

**Comment by erictraut** [comment]
I think this is ready to go. Let me know if you have any other suggested changes.

**Pull request merged**